records/tmpl/templates.go

```diff
@@ -0,0 +1,199 @@
+func DefaultTemplates() []Template { return []Template{"{name}.{framework}"} }
+
+// validPartialLabel checks the validity of a partial label according the RFC, depending on the position
+// of the partial label in the current block between two variables and the position in the whole pattern.
+func validPartialLabel(s string, firstInBlock, lastInBlock, firstInPattern, lastInPattern bool, spec labels.Func) bool {
+	// special case for valid underscores: trim them for spec comparison below
+	labelWithoutValidUnderscore := s
+	if (s[0] == '_' && s != "_") || (s == "_" && !lastInPattern) {
+		labelWithoutValidUnderscore = strings.TrimLeft(s, "_")
+	}
+
+	// prepend or append some character to check for RFC compatibility for non-inner labels.
+	//
+	// But don't do this if there is no variableToken on the left or the right, i.e. these
+	// tokens are the most left or the most right ones in the template.
+	pre := ""
+	post := ""
+	if firstInBlock && !firstInPattern {
+		pre = "a"
+	}
+	if lastInBlock && !lastInPattern {
+		post = "a"
+	}
+
+	escapedLabel := spec(pre + labelWithoutValidUnderscore + post)
+	return pre+labelWithoutValidUnderscore+post == escapedLabel
+}
+
+// addNonVariableTokens splits the given string s into partial labels and adds
+// tokens for them. It accepts labels with "_" depending on whether s is in the
+// left or right most position in the whole pattern.
+func addNonVariableTokens(tokens []token, s string, firstInPattern, lastInPattern bool, spec labels.Func) ([]token, error) {
+	if s == "" {
+		return tokens, nil
+	}
+
+	if s == "." {
+		tokens = append(tokens, separatorToken{})
+		return tokens, nil
+	}
+
+	labels := strings.Split(s, ".")
+	for i, label := range labels {
+		firstInBlock := i == 0            // first partial label in a block between variables or the pattern start/end
+		lastInBlock := i == len(labels)-1 // last partial label in a block between variables or the pattern start/end
+
+		if i != 0 {
+			tokens = append(tokens, separatorToken{})
+		}
+
+		// "" and first or last => . at the left or right of s, skip empty string
+		if label == "" && (firstInBlock || lastInBlock) {
+			continue
+		}
+
+		// "" and not the first or last => consecutive separators
+		if label == "" && !firstInBlock && !lastInBlock {
+			return nil, fmt.Errorf("invalid consecutive separators")
+		}
+
+		if !validPartialLabel(label, i == 0, i == len(labels)-1, firstInPattern, lastInPattern, spec) {
+			return nil, fmt.Errorf("template substring %v is no valid label", label)
+		}
+		tokens = append(tokens, stringToken(label))
+	}
+	return tokens, nil
+}
+
+// Compile compiles a Template to a fast Compiled template.
+func (t Template) Compile(spec labels.Func) (*Compiled, error) {
+	tokens := []token{}
+
+	if string(t) == "" {
+		return nil, fmt.Errorf("invalid empty template")
+	}
+
+	// split template into tokens: strings, separators and {variables}
+	varRE, err := regexp.Compile(`{[\s\w-:]*}`)
```

Comment:
my gut says that a regexp is overkill for the lexer. it's more code to hand-build a lexer but it will probably be more performant and you'll be able to provide better error messages to users when a template has problems.